File: WoofAndroid/presentation/home/src/main/kotlin/com/perrystreet/woof/presentation/home/ui/HomeScreen.kt
package com.perrystreet.woof.presentation.home.ui

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.PreviewParameter
import com.perrystreet.woof.designsystem.atomic.organisms.emptystate.OrgEmptyState
import com.perrystreet.woof.designsystem.atomic.organisms.navbar.OrgBottomNavigationBar
import com.perrystreet.woof.designsystem.atomic.organisms.navbar.OrgBottomNavigationItem
import com.perrystreet.woof.designsystem.atomic.templates.TemplateBottomNavigation
import com.perrystreet.woof.designsystem.preview.PreviewDevices
import com.perrystreet.woof.designsystem.preview.ThemeProvider
import com.perrystreet.woof.designsystem.preview.ThemedScreenPreview
import com.perrystreet.woof.designsystem.theme.ITheme
import com.perrystreet.woof.presentation.home.ui.extensions.HomeTabUIModelExtensions.toBottomNavigationRole
import com.perrystreet.woof.presentation.home.uimodel.HomeTabUIModel
import com.perrystreet.woof.presentation.home.viewmodel.HomeViewModel
import com.perrystreet.woof.resources.R

@Composable
fun HomeScreen(
    state: HomeViewModel.State,
    onTabSelect: (HomeTabUIModel) -> Unit,
    tabContent: @Composable (tab: HomeTabUIModel) -> Unit,
) {
    TemplateBottomNavigation(
        bottomBar = {
            OrgBottomNavigationBar(
                items = state.tabs.map { tab ->
                    OrgBottomNavigationItem(
                        role = tab.toBottomNavigationRole(),
                        isSelected = tab == state.selectedTab,
                        onTap = { onTabSelect(tab) },
                    )
                },
            )
        },
    ) {
        tabContent(state.selectedTab)
    }
}

@PreviewDevices
@Composable
private fun HomeScreenPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        HomeScreen(state = HomeViewModel.State.Initial, onTabSelect = {}) { tab ->
            OrgEmptyState(title = stringResource(R.string.app_name), message = tab.name)
        }
    }
}

File: WoofIOS/SwiftPackages/Presentation/Home/Sources/PresentationHome/UI/HomeScreen.swift
import DesignSystem
import Resources
import SwiftUI

public struct HomeScreen<TabContent: View>: View {
    private let state: HomeViewModel.State
    private let onTabSelect: (HomeTabUIModel) -> Void
    private let tabContent: (HomeTabUIModel) -> TabContent

    public init(
        state: HomeViewModel.State,
        onTabSelect: @escaping (HomeTabUIModel) -> Void,
        @ViewBuilder tabContent: @escaping (HomeTabUIModel) -> TabContent
    ) {
        self.state = state
        self.onTabSelect = onTabSelect
        self.tabContent = tabContent
    }

    public var body: some View {
        TemplateBottomNavigation(
            selectedTab: state.selectedTab,
            items: state.tabs.map { tab in
                let isSelected = tab == state.selectedTab
                return OrgBottomNavigationItem(
                    icon: tab.icon(isSelected: isSelected),
                    label: tab.label,
                    value: tab,
                    isSelected: isSelected,
                    onTap: { onTabSelect(tab) }
                )
            },
            onTabSelect: onTabSelect,
            content: tabContent
        )
    }
}

#Preview("Home") {
    ThemedScreenPreview(theme: WoofTheme.light()) {
        HomeScreen(state: .initial, onTabSelect: { _ in }) { tab in
            OrgEmptyState(title: L10n.appName, message: String(describing: tab))
        }
    }
}
